subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyCollector.java

```diff
@@ -166,6 +178,59 @@ public <D extends Dependency> void add(Provider<? extends D> dependency, Action<
         doAddLazy(dependency, configuration);
     }
 
+
+    private static DependencyConstraint applyConstraintConfiguration(DependencyConstraint dependencyConstraint, @Nullable Action<? super DependencyConstraint> config) {
+        if (config != null) {
+            config.execute(dependencyConstraint);
+        }
+
+        // TODO DependencyConstraint mutation validator?
+
+        return dependencyConstraint;
+    }
+
+    private void doAddConstraintEager(DependencyConstraint dependencyConstraint, @Nullable Action<? super DependencyConstraint> config) {
+        getDependencyConstraints().add(applyConstraintConfiguration(dependencyConstraint, config));
+    }
+
+    private void doAddConstraintLazy(Provider<? extends DependencyConstraint> dependencyConstraint, @Nullable Action<? super DependencyConstraint> config) {
+        Provider<DependencyConstraint> provider = dependencyConstraint.map((Object dep) -> {
+            // Generic failure check (for Groovy which ignores this when dynamic)
+            if (!(dep instanceof DependencyConstraint)) {
+                throw new InvalidUserCodeException(
+                    "Providers of non-DependencyConstraint types ("
+                        + dep.getClass().getName()
+                        + ") are not supported. Create a DependencyConstraint using DependencyConstraintFactory first."
+                );
+            }
+            return applyConstraintConfiguration((DependencyConstraint) dep, config);
+        });
+        getDependencyConstraints().addAll(providerFactory.provider(() -> {
```

Comment:
❓ Seems a little odd to be handling `Provider`s that could return `null` at this point, and handling them in such a way as to make their failure to produce a value not an issue.  I would think we'd want to fail at some point if given a provider producing null, not silently ignore it?